core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -733,13 +733,22 @@ public R get() {
             if (completed.get()) {
                if (exception != null) {
                   throw exception;
+               } else if ((entry = queue.poll()) != null) {
+                  // We check the queue one last time to make sure we didn't have a concurrent queue addition and
+                  // completed iterator
+                  if (consumer != null) {
+                     consumer.accept(entry);
+                  }
+                  return entry;
                }
                return null;
             }
             nextLock.lock();
             try {
                boolean interrupted = false;
-               while ((entry = queue.poll()) == null && !completed.get()) {
+               // We have to poll queue after completed just in case if we completed and sent an item to the
+               // queue and completed iterator that we don't lose the item
```

Comment:
Could you clarify this a bit? I'm still not sure what's going on with the completed iterator :)